research-info.lua

```diff
@@ -0,0 +1,69 @@
+local researchInfo = reqscript("internal/research-info/info")
+
+local function sortResearchDataByPercAndPondered(data_list)
+    table.sort(data_list, function(a, b)
+        if tonumber(a.research_percentage) == tonumber(b.research_percentage) then
+            return a.times_pondered > b.times_pondered
+        end
+        return tonumber(a.research_percentage) > tonumber(b.research_percentage)
+    end)
+
+    return data_list
+end
+
+local function printResearchData(data)
+    if not data then
+        print("No data to display.")
+        return
+    end
+
+    if data.research_points == 0 and data.knowledge_goal == "" and data.times_pondered == 0 then
+        return
+    end
+
+    print("==================")
+    print("Dwarf: " .. data.name)
+    print("\tKnowledge Goal Category: " .. data.knowledge_goal_category)
+    print("\tKnowledge Goal: " .. data.knowledge_goal)
+    print("\tResearch Points: " .. data.research_points .. "/100000")
+    print("\tResearch Percentage: " .. data.research_percentage .. "%")
+    print("\tTimes Pondered: " .. data.times_pondered)
+
+    if #data.research_topics > 0 then
+        print("Knowledge:")
+        for _, topic in ipairs(data.research_topics) do
+            print("\t" .. topic.name .. ": " .. topic.info)
+        end
+    end
+end
+
+--
+local unit  = dfhack.units.getCitizens()
+local selected_unit = dfhack.gui.getSelectedUnit()
+
+if unit then
+    local research_data_list = {}
+    for i, unit in pairs(dfhack.units.getCitizens()) do
+        
+        if selected_unit and selected_unit.id ~= unit.id then
+            goto continue
+        end
+
+        local historical_figure = researchInfo.getHistoricalFigure(unit.id)
```

Comment:
```suggestion
        local historical_figure = df.historical_figure.find(unit.hist_figure_id)
```
There are two observations here: units have a historical figure id field, and the idiomatic way to turn a numeric field into the corresponding structure is to use `df.<structure_name>.find(id)`. This makes functions like your `getHistoricalFigure` basically useless. Please revisit your code and see whether this same pattern can be applied elsewhere.